internal/migrations/changeset_listener.go

```diff
@@ -0,0 +1,323 @@
+package migrations
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/big"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/jpillora/backoff"
+	"github.com/kwilteam/kwil-db/common"
+	"github.com/kwilteam/kwil-db/core/client"
+	"github.com/kwilteam/kwil-db/core/log"
+	"github.com/kwilteam/kwil-db/extensions/listeners"
+)
+
+// Changeset Extension polls the changesets from the old chain  during migration.
+// The changesets received are broadcasted to the eventstore under ChangesetMigrationEvent type for voting.
+// The changesets are then applied to the new chain's database once the resolution is approved.
+const (
+	ListenerName = "migrations"
+	retryDelay   = 1 * time.Second // polling frequency for changesets
+	maxRetries   = 10              // max retries for polling changesets
+)
+
+func init() {
+	// Register the listener with the name "migrations"
+	err := listeners.RegisterListener(ListenerName, Start)
+	if err != nil {
+		panic(err)
+	}
+}
+
+type MigrationConfig struct {
+	// StartHeight is the block height at which the migration started on the old chain.
+	StartHeight int64
+
+	// End height is the block height at which the migration ended on the old chain.
+	EndHeight int64
+
+	// ListenAddress is the address of the kwild server to receive changesets from.
+	ListenAddress string
+}
+
+type migrationListener struct {
+	// Config is the migration configuration
+	config *MigrationConfig
+
+	// Client is the kwild RPC client to the old chain
+	client *client.Client
+
+	// currentHeight is the current block height to be processed
+	currentHeight int64
+
+	// eventStore is the event store to broadcast the changeset migration events
+	eventStore listeners.EventStore
+
+	// logger
+	logger log.SugaredLogger
+}
+
+func Start(ctx context.Context, service *common.Service, eventStore listeners.EventStore) error {
+	// Get the migration config from the service
+	cfg := &MigrationConfig{}
+	migrationCfg, ok := service.ExtensionConfigs[ListenerName]
+	if !ok {
+		service.Logger.Warn("no migration config provided, skipping migration listener")
+		return nil // no migration config, nothing to do
+	}
+
+	err := cfg.ExtractConfig(migrationCfg)
```

Comment:
Is it possible to remove the need to configure a `start_height` and `end_height` by either:

a. passing the genesis params to `*common.Service`
or
b. querying the configured `listen_address`?